Gutenboarding: Allow tabbing in and out of page selector dialog.

yansern opened the pull request:
#### Changes proposed in this Pull Request

This PR allows tabbing in and out of page selector dialog.

The underlying changes includes:
* Disable focus trapping on page selection dialog to allow tabbing in and out. be245bd e491fc4
* Remove use of Portal & DialogBackdrop to allow tabbing out to the elements behind the page selection dialog. fd6bf9e
* Preventing focus on the onboarding block. 659a1fd
* Preventing focus on the design card when on page selection stage. c61e322
* Fix back button disabled when on page selection stage. 2201e07

#### Testing instructions

**Test Case 1: Tabbing-out from the first page selector item**
1. Go to `/gutenboarding`, enter site details & pick a design.
2. When page selector dialog shows up, hit <kbd>Tab</kbd>, it should focus on the first page selector item.
3. Hit <kbd>Shift</kbd>+<kbd>Tab</kbd>, it should focus on the **Create My Site** button.
4. Hit <kbd>Shift</kbd>+<kbd>Tab</kbd> again, it should focus on the **Domain Picker** button.
5. Hit <kbd>Shift</kbd>+<kbd>Tab</kbd> again, it should focus on the **Back Button** button.

**Test Case 2: Tabbing-out from the last page selector item**
1. Go to `/gutenboarding`, enter site details & pick a design.
2. When page selector dialog shows up, click on the last page selector item.
3. Hit <kbd>Tab</kbd>, it should focus on the **Bug Report** button.


#### Screenshot

This screenshot shows the tab ordering of the page. Doing it in reverse (shift-tab) should also follow the same ordering.

![image](https://user-images.githubusercontent.com/1287077/74144741-93668500-4bfd-11ea-8fe6-d8a8289284ff.png)

Fixes #38664 #38635

ockham commented on client/landing/gutenboarding/gutenboard.tsx:
```diff
@@ -33,6 +33,9 @@ export function Gutenboard() {
 			break;
+		case Step.PageSelection:
+			prev = Step.DesignSelection;
+			break;
```
Ah, good point, forget that when we were adding the Page Selector route (#39210)! :sweat_smile:

yansern replied:
Yes totally missed that one.

ockham commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -152,21 +152,14 @@ const DesignSelector: FunctionComponent< Props > = ( { showPageSelector = false
-					className="design-selector__page-layout-backdrop"
-				/>
-			</Portal>
-
```
Curious if there was a reason why we had this in the first place -- basically to make sure we're not missing anything that would introduce any regressions here? @sirreal @razvanpapadopol

sirreal replied:
This seemed necessary for some of the modal-like behavior. I believe hide on click outside didn't work properly without it.

yansern replied:
Yes, this is for hide on click outside. But there is no more hide on click outside, as we now allow them to click outside to buttons like the back button and the domain picker button.

ockham commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -152,21 +152,14 @@ const DesignSelector: FunctionComponent< Props > = ( { showPageSelector = false
 				} }
 				aria-labelledby="page-layout-selector__title"
-				hideOnClickOutside
+				hideOnClickOutside={ false }
```
Do we need this to fix tabbing? It seems kinda convenient to be able to leave the page selector by clicking outside of it?

simison replied:
>It seems kinda convenient to be able to leave the page selector by clicking outside of it?

I'm not sure about that. Imagine clicking at the white page area outside the design selector grid and be thrown at intent gather step. Since it's not visually resembling modal as much, I think we're better with explicit back buttons, either contextual in the content area or in the header.

sirreal replied:
We need to be able to click outside on buttons in the header and whatnot. This is probably necessary.

One note, `hideOnClickOutside` sets the prop to boolean `true`. Assuming that's not the default, we can remove the prop rather than explicitly setting it to `false`:

```suggestion
```

simison replied:
Aight, sounds good! 👍

yansern replied:
> Do we need this to fix tabbing? It seems kinda convenient to be able to leave the page selector by clicking outside of it?

Yes indeed, because now that back button, domain picker button and create new site button is clickable/tabbable, the `hideOnClickOutside` behaviour is removed.

> One note, hideOnClickOutside sets the prop to boolean true. Assuming that's not the default, we can remove the prop rather than explicitly setting it to false.

Yes, that's right. `hideOnClickOutside` has a default value of `true`, that's why this has to be set explicitly to `false`.

ockham commented on client/landing/gutenboarding/onboarding-block/design-selector/style.scss:
```diff
@@ -1,9 +1,10 @@
 
 $design-selector-max-width: 1100px;
-$design-selector-selection-space: 285px;
+$design-selector-selection-space: 285px + $header-height;
```
I don't quite understand the styling changes in here -- are the required for the tabbing fix? :slightly_smiling_face:

yansern replied:
`$design-selector-selection-space` refers to the space that allows you to peek through the design selector UI when you are on page selection.

With the page selector modal being a child element of the design selector, the selection space has to take account of the header height, otherwise, the content of the design selector becomes clipped (see screenshot).

![image](https://user-images.githubusercontent.com/1287077/74738661-b0cbcc80-5257-11ea-9f71-866146f85e97.png)

ockham commented on client/landing/gutenboarding/onboarding-block/index.ts:
```diff
@@ -32,4 +32,7 @@ export const settings: BlockConfiguration< Attributes > = {
 	save: () => null,
+	getEditWrapperProps() {
+		return { tabIndex: -1 };
+	},
```
Nice :+1:

yansern replied:
Thanks!

simison commented:
I just wanted to say that these are some of the best testing instructions I've seen in a while. 😮

yansern commented:
> There's a styling issue. When the page selector is added, we now get horizontal scrolling:

@sirreal Good catch with the horizontal scrolling bug. Let me see if I can replicate this and I'll get back to you.

> I'd recommend adding the Portal and restoring the 100% width, but I'm not sure how that will affect the other changes here.

Actually, not using Portal one of the key to make tabbing in & out work. Using Portal requires additional work like tab focus redirection, and creating references to other buttons where to redirect the focus to.

Going back to Portal is just going to be one big headache all over again, so I would rather move forward to see if horizontal scrolling issue can be fixed through CSS (or maybe with some JS help).

yansern commented:
> @sirreal Good catch with the horizontal scrolling bug. Let me see if I can replicate this and I'll get back to you.

@sirreal I am unable to replicate the horizontal scroll bug on both Chrome & Firefox. I also tried using different zoom levels. We might need to do a screen sharing session when you are free.

sirreal commented:
> @sirreal I am unable to replicate the horizontal scroll bug on both Chrome & Firefox. I also tried using different zoom levels and varying browser width. We might need to do a screen sharing session when you are free.

Scrollbars can be funny. Do you have this setting enabled? I'd recommend `always`:

![Screen Shot 2020-02-19 at 10 28 07](https://user-images.githubusercontent.com/841763/74820677-b92d1180-5302-11ea-80c5-8fd061a79cab.png)

yansern commented:
@sirreal You might not like the CSS hack I've done to make the horizontal scrollbar go away here https://github.com/Automattic/wp-calypso/pull/39336/commits/736729b26237058b0e433c4ac25db9e3fc4b2613.

Basically, what it does is reducing the width which leaves a gap on the left and right side of the page selector body, and then fill that gap using box-shadow which renders outside of the box-model hence it fills up the gap without occupying any width.

Hack feels a bit dirty but I think it is fine. There's going to be a redesign happening anyway!

yansern commented:
Rebased.

yansern commented:
Update: After rebasing, there is a hidden element being focused on (a hidden button on block toolbar) when tabbing between 3 & 4 (create site button & first page layout selector). This will be fixed in another PR.

p-jackson commented:
Just a couple of points:
1. I'm not sure pressing Esc should still close the page selector, but I could be convinced either way
2. The tabbing looks great in Chrome, but doesn't seem to be working in FF for some reason. The `tabIndex=-1` for the design card doesn't seem to be doing the trick

**Tabbing in Chrome**
![chrometab](https://user-images.githubusercontent.com/1500769/74897518-34262300-53fc-11ea-8202-aa251d52872c.gif)

**Tabbing in Firefox**
![fftab](https://user-images.githubusercontent.com/1500769/74897532-3be5c780-53fc-11ea-8198-2df3488a8566.gif)

ockham commented:
> Update: After rebasing, there is a hidden element being focused on (a hidden button on block toolbar) when tabbing between 3 & 4 (create site button & first page layout selector). This will be fixed in another PR.
> 
> I decided to roll this fix in this PR. [444cc85](https://github.com/Automattic/wp-calypso/commit/444cc85f7dc0bf7d9d025ce00162c34961476460)
> 
> Also fixes #38635.

Overall, since this addresses a separate issue, I think it'd make sense to file a separate PR for this change, to keep our PRs atomic. (This would allow us to e.g. merge the style fix before the tabbing related fixes, in case the latter turn out to need more changes.)

I guess it's fine to keep it in this PR for now -- just a note for future PRs :slightly_smiling_face:

yansern commented:
> The tabbing looks great in Chrome, but doesn't seem to be working in FF for some reason. The tabIndex=-1 for the design card doesn't seem to be doing the trick

@p-jackson Actually it's not the design card, it seems that Firefox treat `.design-selector__grid-container` as tabbable, I don't see any reason why it should be tabbable, could be a quirk with Firefox. Added `tabindex=1` on this element as well. https://github.com/Automattic/wp-calypso/pull/39336/commits/9247566a4ac62a03627ac5c0d2a85bfde6a4e3c5

yansern commented on client/landing/gutenboarding/gutenboard.tsx:
```diff
@@ -62,7 +65,13 @@ export function Gutenboard() {
+									__experimentalUIParts={ {
+										hasPopover: false,
+										hasSelectedUI: false,
+									} }
```
Thanks @p-jackson. This is really helpful as I have zero TypeScript experience! Added here: https://github.com/Automattic/wp-calypso/pull/39336/commits/f4c65c0eeb447ee35aa87beb0a1cb17ca9b6d3a9

yansern commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -129,6 +128,7 @@ const DesignSelector: FunctionComponent< Props > = ( { showPageSelector = false
 											visibility: showPageSelector ? 'hidden' : undefined,
 									  }
 							}
+							tabIndex={ showPageSelector ? -1 : 0 }
```
Hmm, thanks for the suggestion but I think I will leave this as it is as I feel this introduces unnecessary abstraction... for something as trivial as tabIndex. :)

p-jackson replied:
Fair point

yansern commented on client/landing/gutenboarding/gutenboard.tsx:
```diff
@@ -62,7 +65,13 @@ export function Gutenboard() {
 							>
-								<BlockList className="gutenboarding-block-list" />
+								<BlockList
+									className="gutenboarding-block-list"
```
Removed here. https://github.com/Automattic/wp-calypso/pull/39336/commits/43792873c66aed506e799742243a90140d953cb9

yansern commented:
> This would allow us to e.g. merge the style fix before the tabbing related fixes, in case the latter turn out to need more changes.

You've got a point there @ockham! Thanks!

razvanpapadopol commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -112,6 +111,7 @@ const DesignSelector: FunctionComponent< Props > = ( { showPageSelector = false
 				className={ classnames( 'design-selector__grid-container', {
 					'is-page-selector-open': showPageSelector,
 				} ) }
+				tabIndex="-1"
```
To fix typescript check failing:
```suggestion
				tabIndex={-1}
```

yansern replied:
Thank you @razvanpapadopol !

sirreal commented on client/landing/gutenboarding/gutenboard.tsx:
```diff
@@ -62,7 +74,12 @@ export function Gutenboard() {
+									__experimentalUIParts={ {
+										hasPopover: false,
+										hasSelectedUI: false,
+									} }
```
The smaller and more focused we can keep PRs, the better. This would probably have made a nice, small, contained PR by itself. This PR has a lot going on and adding more tends to slow things down.

No problem here, just something to keep in mind 👍

yansern replied:
It did slow things down! Lesson learned!

sirreal commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -112,6 +111,7 @@ const DesignSelector: FunctionComponent< Props > = ( { showPageSelector = false
 				className={ classnames( 'design-selector__grid-container', {
 					'is-page-selector-open': showPageSelector,
 				} ) }
+				tabIndex={-1}
```
Should be surrounding space here - did prettier not handle this?

```suggestion
				tabIndex={ -1 }
```

yansern replied:
It was applied through GitHub's commit suggestion button, no prettier ran there. Better to do it locally next time!

sirreal commented on client/landing/gutenboarding/onboarding-block/design-selector/design-card.tsx:
```diff
@@ -19,8 +19,15 @@ interface Props {
+	dialogId,
+	onClick,
+	style,
+	tabIndex = 0,
```
Is `tabindex="0"` on an interactive element (like the `button` we'll render here) redundant? If so, we can omit this and don't need to worry about setting the attribute at all.

yansern replied:
The parent component is deciding whether to set `tabIndex` to `0` or `-1` based on the visibility of the page selector that's why it's here.

sirreal replied:
This is a minor point, but I don't think I was clear to I'll elaborate. This is an opportunity for the parent to pass _any_ `tabIndex` (not just `0` or `1`), which will be passed to the `Card`.

My question was about the default `0`. If the `tabIndex` is not provided by the parent, `<button tabindex="0"/>` will have the same tabbing behavior as `<button />` without the attribute, right? In that case, we don't need to pass `0` at all:

```suggestion
	tabIndex,
```

yansern commented:
@razvanpapadopol Thanks! You have quite the attention for detail!

> entering animation is too quick; it doesn't seem to be starting from the bottom anymore
> exit animation seems to be broken; it starts very slowly but then jumps to the end

Fixed in https://github.com/Automattic/wp-calypso/pull/39336/commits/fcd8ae37d1573229b60a9874082c79e259de5bfb by moving the wrapper to the outer element. Turns out to be an easier fix than I thought.

> selected design card displayed at the top is still hoverable with mouse

Fixed here. https://github.com/Automattic/wp-calypso/pull/39336/commits/011a2c60ee0cb9c518048c57bb30ba0982e4440b

yansern commented on client/landing/gutenboarding/onboarding-block/design-selector/style.scss:
```diff
@@ -50,10 +53,16 @@ $design-selector-selection-space: 285px;
+		box-shadow: $white 0 $page-selector-shadow-size 0 $page-selector-width-offset,
+			$dark-gray-300 0 ( $page-selector-shadow-size - $page-selector-width-offset )
+				$page-selector-shadow-size;
+	}
```
Page layout selector container is visible without content when on design selector stage. Setting `visibility: hidden` messes with animation (as the animation had no starting point to animate when visibility is hidden). Quick fix was to do it this way.

yansern commented on client/landing/gutenboarding/onboarding-block/design-selector/index.tsx:
```diff
@@ -155,29 +157,25 @@ const DesignSelector: FunctionComponent< Props > = ( { showPageSelector = false
-				</animated.div>
-			</Dialog>
+				</Dialog>
+			</animated.div>
```
These entire block is to move the `<animated.div>` wrapper out from the `<PageLayoutSelector>` into `<Dialog>` to fix the enter/exit animation regression.

yansern commented:
> It also looks like this needs a rebase.
@sirreal Rebased!

yansern commented:
@p-jackson Will follow-up in a separate issue/PR so I can get this merged! :)
